File: brokers/fubon-stock-cpp/src/latency_measurement.cpp
#include "latency_measurement.h"

#include <curl/curl.h>

using namespace std::chrono;

void LatencyMeasurement::loadConfig(const std::string& filename) {
  try {
    config_table_ = toml::parse_file(filename);
  } catch (const toml::parse_error& err) {
    std::cerr << "ERROR: Failed to parse " << filename << ": " << err.what()
              << std::endl;
    exit(1);
  }
}

LatencyMeasurement::LatencyMeasurement() {
  loadConfig();
  sdk_ = std::make_unique<fubon::FubonSDK>();
  callback_ = std::make_shared<NoOpCallback>();
  sdk_->register_callback(callback_);
}

LatencyMeasurement::~LatencyMeasurement() {
  if (sdk_ && logged_in_) {
    sdk_->logout();
  }
}

bool LatencyMeasurement::initialize() {
  auto user_id = getConfig<std::string>("user", "user_id");
  auto password = getConfig<std::string>("user", "password");
  auto cert_path = getConfig<std::string>("user", "pfx_filepath");
  auto cert_pass = getConfig<std::string>("user", "pfx_password");

  auto resp = sdk_->login(user_id, password, cert_path, cert_pass);
  if (!resp.is_success) {
    std::cerr << "Failed to connect to broker: "
              << resp.message.value_or("(no message)") << std::endl;
    return false;
  }
  if (!resp.data.has_value() || resp.data.value().empty()) {
    std::cerr << "Failed to connect to broker: no accounts" << std::endl;
    return false;
  }

  account_ = resp.data.value()[0];
  logged_in_ = true;

  std::cout << "Successfully connected and authenticated with broker"
            << std::endl;
  return true;
}

void LatencyMeasurement::submitOrder() {
  using namespace fubon_parsers;

  order_counter_++;
  auto now = std::chrono::floor<seconds>(system_clock::now());
  std::chrono::zoned_time local{std::chrono::current_zone(), now};
  std::string order_id =
      std::format("test-{:%Y%m%d%H%M%S}-{}", local, order_counter_);

  fubon::Order order{
      parseSide(getConfig<std::string>("order", "side")),
      getConfig<std::string>("order", "symbol"),
      std::format("{:g}", getConfig<double>("order", "price")),
      getConfig<int64_t>("order", "quantity"),
      parseMarketType(getConfig<std::string>("order", "market_type")),
      parsePriceType(getConfig<std::string>("order", "price_type")),
      parseTimeInForce(getConfig<std::string>("order", "time_in_force")),
      parseOrderType(getConfig<std::string>("order", "order_type")),
      order_id,
  };

  auto t1 = high_resolution_clock::now();
  auto resp = sdk_->stock->place_order(account_, order, std::nullopt);
  auto t2 = high_resolution_clock::now();

  if (!resp.is_success || !resp.data.has_value()) {
    std::cerr << "Order submission failed: "
              << resp.message.value_or("(no message)") << std::endl;
    return;
  }

  measured_latency_ms_ =
      duration_cast<microseconds>(t2 - t1).count() / 1000.0;

  std::cout << "Order submitted successfully!" << std::endl;
  std::cout << "Round-trip latency: " << measured_latency_ms_ << " ms"
            << std::endl;

  std::this_thread::sleep_for(std::chrono::seconds(1));

  std::cout << "Cancelling order..." << std::endl;
  auto cancel_resp =
      sdk_->stock->cancel_order(account_, resp.data.value(), std::nullopt);

  if (cancel_resp.is_success) {
    std::cout << "Order cancelled successfully!" << std::endl;
    std::cout << "Order ID: "
              << resp.data.value().order_no.value_or("(none)") << std::endl;
    std::cout << "Order lifecycle completed (submit -> cancel)" << std::endl;
    sendLatencyReport();
  } else {
    std::cerr << "Order cancellation failed: "
              << cancel_resp.message.value_or("(no message)") << std::endl;
  }
}

void LatencyMeasurement::sendLatencyReport() {
  std::string api_url = getConfig<std::string>("api", "url");
  std::string broker = getConfig<std::string>("api", "broker_name");
  std::string symbol = getConfig<std::string>("order", "symbol");
  std::string side_raw = getConfig<std::string>("order", "side");
  std::string side =
      (!side_raw.empty() && (side_raw[0] == 'B' || side_raw[0] == 'b')) ? "B"
                                                                       : "S";
  double price = getConfig<double>("order", "price");
  int64_t volume = getConfig<int64_t>("order", "quantity");

  auto now_s = std::chrono::floor<seconds>(system_clock::now());
  std::string timestamp = std::format("{:%Y-%m-%dT%H:%M:%SZ}", now_s);

  std::string body = std::format(
      R"({{"timestamp":"{}","broker":"{}","latency_ms":{:.3f},)"
      R"("symbol":"{}","side":"{}","price":{},"volume":{}}})",
      timestamp, broker, measured_latency_ms_, symbol, side, price, volume);

  CURL* curl = curl_easy_init();
  if (!curl) {
    std::cerr << "Failed to send latency report: curl_easy_init failed"
              << std::endl;
    return;
  }

  struct curl_slist* headers =
      curl_slist_append(nullptr, "Content-Type: application/json");

  curl_easy_setopt(curl, CURLOPT_URL, api_url.c_str());
  curl_easy_setopt(curl, CURLOPT_POST, 1L);
  curl_easy_setopt(curl, CURLOPT_POSTFIELDS, body.c_str());
  curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, (long)body.size());
  curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
  curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);
  curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1L);
  curl_easy_setopt(
      curl, CURLOPT_WRITEFUNCTION,
      +[](char*, size_t s, size_t n, void*) -> size_t { return s * n; });

  CURLcode res = curl_easy_perform(curl);
  long status_code = 0;
  if (res == CURLE_OK) {
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &status_code);
  }

  curl_slist_free_all(headers);
  curl_easy_cleanup(curl);

  if (res != CURLE_OK || status_code < 200 || status_code >= 300) {
    std::cerr << "Failed to send latency report: " << curl_easy_strerror(res)
              << " (HTTP " << status_code << ")" << std::endl;
  }
}

void LatencyMeasurement::runLatencyTest() {
  std::string symbol = getConfig<std::string>("order", "symbol");
  double price = getConfig<double>("order", "price");
  int64_t quantity = getConfig<int64_t>("order", "quantity");
  std::string side_str = getConfig<std::string>("order", "side");
  int interval = getConfig<int>("trading_hours", "interval_seconds");
  std::string start_time_str =
      getConfig<std::string>("trading_hours", "start_time");
  std::string end_time_str =
      getConfig<std::string>("trading_hours", "end_time");

  int start_time =
      std::stoi(start_time_str.substr(0, 2) + start_time_str.substr(3, 2));
  int end_time =
      std::stoi(end_time_str.substr(0, 2) + end_time_str.substr(3, 2));

  std::cout << "Trading: " << symbol << " " << side_str << " " << price << " x"
            << quantity << " every " << interval << "s during "
            << start_time_str << "-" << end_time_str << std::endl;

  while (true) {
    auto now = system_clock::now();
    auto utc8_now = now + std::chrono::hours(8);
    auto utc8_dp = std::chrono::floor<std::chrono::days>(utc8_now);
    std::chrono::weekday wd{utc8_dp};
    std::chrono::hh_mm_ss hms{utc8_now - utc8_dp};

    int current_time = hms.hours().count() * 100 + hms.minutes().count();
    bool is_weekday =
        (wd != std::chrono::Saturday && wd != std::chrono::Sunday);
    bool is_trading_time =
        (current_time >= start_time && current_time <= end_time);

    if (!is_weekday || !is_trading_time) {
      std::this_thread::sleep_for(std::chrono::minutes(1));
      continue;
    }

    std::cout << "\n--- Order #" << order_counter_ + 1 << " ---" << std::endl;
    submitOrder();
    std::this_thread::sleep_for(std::chrono::seconds(interval));
  }
}

File: brokers/fubon-stock-python/latency_measurement.py
import os
import time
import requests
import toml
import logging
import traceback
from datetime import datetime, timezone, timedelta

import fubon_neo
import fubon_neo.constant
from fubon_neo.sdk import FubonSDK, Order

logging.basicConfig(level=logging.INFO, format="%(message)s")
logger = logging.getLogger(__name__)

TAIWAN_TZ = timezone(timedelta(hours=8))


class LatencyMeasurement:
    """Measures order submission latency for Fubon stock trading."""

    def __init__(self, config_path="config.toml"):
        self.config = toml.load(config_path)
        self.validate_config()

        self.sdk = FubonSDK()
        self.account = None
        self.measured_latency_ms = 0.0

    def validate_config(self):
        required = {
            "user": ["user_id", "password", "pfx_filepath", "pfx_password"],
            "order": [
                "symbol",
                "price",
                "quantity",
                "action",
                "market_type",
                "price_type",
                "time_in_force",
                "order_type",
            ],
            "trading_hours": ["start_time", "end_time", "interval_seconds"],
            "api": ["url", "broker_name"],
        }
        for section, fields in required.items():
            if section not in self.config:
                raise ValueError(f"Missing config section: [{section}]")
            for field in fields:
                if field not in self.config[section]:
                    raise ValueError(f"Missing required config: [{section}].{field}")

        if self.config["order"]["price"] <= 0:
            raise ValueError("Price must be positive")
        if self.config["order"]["quantity"] <= 0:
            raise ValueError("Quantity must be positive")
        if self.config["trading_hours"]["interval_seconds"] <= 0:
            raise ValueError("Interval must be positive")

    def login(self):
        try:
            user = self.config["user"]
            resp = self.sdk.login(
                user["user_id"],
                user["password"],
                user["pfx_filepath"],
                user["pfx_password"],
            )
            if not resp.is_success or not resp.data:
                logger.error(
                    f"Failed to connect to broker: {resp.message or '(no message)'}"
                )
                return False

            self.account = resp.data[0]
            logger.info("Successfully connected and authenticated with broker")
            return True

        except Exception as e:
            logger.error(f"Failed to connect to broker: {e}")
            traceback.print_exc()
            return False

    def submit_order(self, symbol, action, price, quantity):
        try:
            order = Order(
                buy_sell=getattr(fubon_neo.constant.BSAction, action),
                symbol=symbol,
                price=str(price),
                quantity=quantity,
                market_type=getattr(
                    fubon_neo.constant.MarketType,
                    self.config["order"]["market_type"],
                ),
                price_type=getattr(
                    fubon_neo.constant.PriceType,
                    self.config["order"]["price_type"],
                ),
                time_in_force=getattr(
                    fubon_neo.constant.TimeInForce,
                    self.config["order"]["time_in_force"],
                ),
                order_type=getattr(
                    fubon_neo.constant.OrderType,
                    self.config["order"]["order_type"],
                ),
            )

            t1 = time.perf_counter()
            resp = self.sdk.stock.place_order(self.account, order)
            t2 = time.perf_counter()

            if not resp.is_success or resp.data is None:
                logger.error(
                    f"Order submission failed: {resp.message or '(no message)'}"
                )
                return

            self.measured_latency_ms = (t2 - t1) * 1000

            logger.info("Order submitted successfully!")
            logger.info(f"Round-trip latency: {self.measured_latency_ms:.2f} ms")

            side = "B" if action.lower() == "buy" else "S"
            self.send_latency_report(
                symbol=symbol,
                side=side,
                price=price,
                volume=quantity,
                latency_ms=self.measured_latency_ms,
            )

            time.sleep(1)

            logger.info("Cancelling order...")
            cancel_resp = self.sdk.stock.cancel_order(self.account, resp.data)

            if cancel_resp.is_success:
                order_no = cancel_resp.data.order_no if cancel_resp.data else "(none)"
                logger.info("Order cancelled successfully!")
                logger.info(f"Order ID: {order_no}")
                logger.info("Order lifecycle completed (submit -> cancel)")
            else:
                logger.error(
                    f"Order cancellation failed: {cancel_resp.message or '(no message)'}"
                )

        except Exception as e:
            logger.error(f"Order submission failed: {e}")
            traceback.print_exc()

    def send_latency_report(self, symbol, side, price, volume, latency_ms):
        try:
            api_url = self.config["api"]["url"]
            broker_name = self.config["api"]["broker_name"]

            data = {
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "broker": broker_name,
                "latency_ms": latency_ms,
                "symbol": symbol,
                "side": side,
                "price": price,
                "volume": volume,
            }

            response = requests.post(api_url, json=data, timeout=5)

            if not (200 <= response.status_code < 300):
                logger.warning(
                    f"Failed to send latency report: HTTP {response.status_code}"
                )

        except requests.exceptions.Timeout:
            logger.error("Timeout sending latency report")
        except requests.exceptions.RequestException as e:
            logger.error(f"Error sending latency report: {e}")
        except Exception as e:
            logger.error(f"Unexpected error sending latency report: {e}")

    def get_current_time_hhmm(self):
        taiwan_time = datetime.now(TAIWAN_TZ)
        return taiwan_time.hour * 100 + taiwan_time.minute

    def is_weekday(self):
        return datetime.now(TAIWAN_TZ).weekday() < 5

    def run_latency_test(self):
        symbol = self.config["order"]["symbol"]
        price = self.config["order"]["price"]
        quantity = self.config["order"]["quantity"]
        action = self.config["order"]["action"]
        interval = self.config["trading_hours"]["interval_seconds"]

        start_time_str = self.config["trading_hours"]["start_time"]
        end_time_str = self.config["trading_hours"]["end_time"]
        start_time = int(start_time_str.replace(":", ""))
        end_time = int(end_time_str.replace(":", ""))

        side = "B" if action.lower() == "buy" else "S"
        logger.info(
            f"Trading: {symbol} {side} {price} x{quantity} every {interval}s "
            f"during {start_time_str}-{end_time_str}"
        )

        order_count = 0

        while True:
            try:
                current_time = self.get_current_time_hhmm()
                is_trading_time = start_time <= current_time <= end_time

                if not self.is_weekday() or not is_trading_time:
                    time.sleep(60)
                    continue

                order_count += 1
                logger.info(f"\n--- Order #{order_count} ---")

                self.submit_order(symbol, action, price, quantity)
                time.sleep(interval)

            except KeyboardInterrupt:
                break
            except Exception as e:
                logger.error(f"Error in test loop: {e}")
                traceback.print_exc()
                time.sleep(interval)


def main():
    latency_test = LatencyMeasurement()

    if not latency_test.login():
        return 1

    time.sleep(2)

    try:
        latency_test.run_latency_test()
    except KeyboardInterrupt:
        pass
    except Exception as e:
        logger.error(f"Fatal error: {e}")
        traceback.print_exc()
        return 1

    return 0


if __name__ == "__main__":
    rc = main()
    os._exit(rc)
